sync3/handler/connstate.go

```diff
@@ -543,3 +556,25 @@ func (s *ConnState) OnRoomUpdate(ctx context.Context, up caches.RoomUpdate) {
 		logger.Warn().Str("room_id", up.RoomID()).Msg("OnRoomUpdate unknown update type")
 	}
 }
+
+// clampSliceRangeToListSize helps us send client-friend SYNC and INVALIDATE ranges.
+//
+// Suppose the client asks for a window on positions [10, 19]. If the list
+// has exactly 12 rooms, the window will see 3: those at positions 10, 11
+// and 12. In this situation, it is helpful for clients to be given the
+// "effective" range [10, 12] rather than the "conceptual" range [10, 19].
+//
+// The "full" room list occupies positions [0, totalRooms - 1]. If the given range r
+// does not overlap the full room list, return nil. Otherwise, return the intersection
+// of r with the full room list.
+func clampSliceRangeToListSize(r [2]int64, totalRooms int64) []int64 {
```

Comment:
This signature could be nicer if we just pass through `r` unaltered in cases where it doesn't reach / is beyond `totalRooms` and then clamp if the range is within the index position. That way, rather than callers being forced to `if nil` check, they can just in-line it e.g:
```
Range: clampSliceRangeToListSize(removedRanges[i][:], totalRooms)
```

I do not expect this function to perform validation on the range, as that should have already happened.

Reply:
> This signature could be nicer if we just pass through r unaltered in cases where it doesn't reach / is beyond totalRooms and then clamp if the range is within the index position.

I can see that this works where the range doesn't reach the totalRooms (line 577). But when the start of the range is already beyond total rooms (572), I don't want to tell the client about this range at all---hence the nil return value and `continue`s. But perhaps by

> I do not expect this function to perform validation on the range, as that should have already happened.

you mean that the start of the range `r[0]` should never be beyond `totalRooms`? If so I don't follow, since the start of the range comes from the client.

Reply:
Let's write a test to see if this is actually possible in practice.

Reply:
The test

```go
// Tests the scenario described at
// https://github.com/matrix-org/sliding-sync/pull/58#discussion_r1159850458
func TestRangesThatExceedTotalRooms(t *testing.T) {
	alice := registerNewUser(t)

	t.Log("Alice makes three public rooms.")
	room0 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "A"})
	room1 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "B"})
	alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "C"})

	t.Log("Alice initial syncs, requesting room ranges [0, 1] and [8, 9]")
	syncRes := alice.SlidingSync(t, sync3.Request{
		Lists: map[string]sync3.RequestList{
			"a": {
				Sort:   []string{sync3.SortByName},
				Ranges: sync3.SliceRanges{{0, 1}, {8, 9}},
			},
		},
	})

	t.Log("Alice should only see rooms 0 and 1 in the sync response.")
	m.MatchResponse(
		t,
		syncRes,
		m.MatchList(
			"a",
			m.MatchV3Count(3),
			m.MatchV3Ops(
				m.MatchV3SyncOp(0, 1, []string{room0, room1}),
			),
		),
	)
}
```

passes on main (690790e46d8d477ede3b5a1f21f27726d3fb7a06). So this doesn't seem to happen in practice, presumably due to the continue in

 https://github.com/matrix-org/sliding-sync/blob/main/sync3/handler/connstate.go#L276-L293

Will amend the clamping function as above.

Reply:
> Will amend the clamping function as above.

Having done this, we are now seeing assertion failures showing that this logic was necessary, e.g. https://github.com/matrix-org/sliding-sync/actions/runs/4667271913/jobs/8262861981?pr=58#step:7:394

I can either
- go back to returning a possibly nil slice
- fixup the logic so that we never try to clamp an out-of-bounds range.

Any preference?

Reply:
> fixup the logic so that we never try to clamp an out-of-bounds range.

Have updated the PR to do this.